prov/verbs/src/verbs_eq.c

```diff
@@ -797,32 +878,42 @@ fi_ibv_eq_read(struct fid_eq *eq_fid, uint32_t *event,
 	return -FI_EAGAIN;
 }
 
+static ssize_t
+fi_ibv_eq_read(struct fid_eq *eq_fid, uint32_t *event, void *buf,
+               size_t len, uint64_t flags)
+{
+	return fi_ibv_eq_read_internal(eq_fid, NULL, event, buf, len, flags);
+}
+
 static ssize_t
 fi_ibv_eq_sread(struct fid_eq *eq_fid, uint32_t *event,
 		void *buf, size_t len, int timeout, uint64_t flags)
 {
 	struct fi_ibv_eq *eq;
-	struct epoll_event events[2];
+	struct epoll_event events;
+	struct fi_ibv_domain *domain = NULL;
 	ssize_t ret;
 
 	eq = container_of(eq_fid, struct fi_ibv_eq, eq_fid.fid);
 
 	while (1) {
-		ret = fi_ibv_eq_read(eq_fid, event, buf, len, flags);
+		ret = fi_ibv_eq_read_internal(eq_fid, domain, event, buf, len, flags);
 		if (ret && (ret != -FI_EAGAIN))
 			return ret;
 
-		ret = epoll_wait(eq->epfd, events, 2, timeout);
+		ret = epoll_wait(eq->epfd, &events, 1, timeout);
```

Comment:
eq_read_internal now calls epoll_wait. We should be able to find a way to remove calling it here, but maybe it gets messy?  It's just a little confusing how domain must be NULL before the first time we call read_internal, then it gets set, so that we skip calling epoll_wait internally.

Reply:
You are right and this part of the code is confusing. I revisited it, let me know what you think of the new version.